src/Compilers/CSharp/Portable/Parser/LanguageParser_InterpolatedString.cs

```diff
@@ -39,98 +40,234 @@ private ExpressionSyntax ParseInterpolatedStringToken()
+                // Errors are positioned relative to the start of the token that was lexed.  Specifically relative to
+                // the starting `$` or `@`.  However, when placed on a node like this, it will be relative to the node's
+                // full start.  So we have to adjust the diagnostics taking that into account.
+                result = result.WithDiagnosticsGreen(MoveDiagnostics(new[] { error }, originalToken.GetLeadingTrivia()?.FullWidth ?? 0));
             }
 
             Debug.Assert(originalToken.ToFullString() == result.ToFullString()); // yield from text equals yield from node
             return result;
 
-            void rescanInterpolation(out Lexer.InterpolatedStringKind kind, out SyntaxDiagnosticInfo error, out Range openQuoteRange, ArrayBuilder<Lexer.Interpolation> interpolations, out Range closeQuoteRange)
+            void rescanInterpolation(out Lexer.InterpolatedStringKind kind, out SyntaxDiagnosticInfo? error, out Range openQuoteRange, ArrayBuilder<Lexer.Interpolation> interpolations, out Range closeQuoteRange)
             {
                 using var tempLexer = new Lexer(SourceText.From(originalText), this.Options, allowPreprocessorDirectives: false);
                 var info = default(Lexer.TokenInfo);
                 tempLexer.ScanInterpolatedStringLiteralTop(ref info, out error, out kind, out openQuoteRange, interpolations, out closeQuoteRange);
             }
 
-            SyntaxToken getOpenQuote(Range openQuoteRange)
+            SyntaxToken getOpenQuote()
             {
-                var openQuoteText = originalText[openQuoteRange];
                 return SyntaxFactory.Token(
                     originalToken.GetLeadingTrivia(),
-                    kind is Lexer.InterpolatedStringKind.Verbatim ? SyntaxKind.InterpolatedVerbatimStringStartToken : SyntaxKind.InterpolatedStringStartToken,
-                    openQuoteText, openQuoteText, trailing: null);
+                    kind switch
+                    {
+                        Lexer.InterpolatedStringKind.Normal => SyntaxKind.InterpolatedStringStartToken,
+                        Lexer.InterpolatedStringKind.Verbatim => SyntaxKind.InterpolatedVerbatimStringStartToken,
+                        Lexer.InterpolatedStringKind.SingleLineRaw => SyntaxKind.InterpolatedSingleLineRawStringStartToken,
+                        Lexer.InterpolatedStringKind.MultiLineRaw => SyntaxKind.InterpolatedMultiLineRawStringStartToken,
+                        _ => throw ExceptionUtilities.UnexpectedValue(kind),
+                    },
+                    originalText[openQuoteRange],
+                    trailing: null);
             }
 
-            CodeAnalysis.Syntax.InternalSyntax.SyntaxList<InterpolatedStringContentSyntax> getContent(ArrayBuilder<Lexer.Interpolation> interpolations)
+            CodeAnalysis.Syntax.InternalSyntax.SyntaxList<InterpolatedStringContentSyntax> getContent(ReadOnlySpan<char> originalTextSpan)
```

Comment:
I'm a little confused by the mix of capture and argument passing (why choose one versus the other for different locals from containing method). 
If it's not too ugly, consider switching to static local functions with explicit arguments for everything.

Reply:
yeah... i'm on the fence.  i def think mixing is odd, so i'll stop doing that.   however, as static, you end up needing to pass quite a bit.  let me try passing nothing, and you tell me if it's too confusing.

Reply:
oh... i have to pass this in as it's a ref struct, so it can't be captured by a local function :(

Reply:
You could consider passing in the original string reference, then making a span to it locally. AsSpan is pretty cheap, but probably not worth it.

Reply:
that's an interesting idea... let me play around and see if i like it at all.  note: i wish i could just capture ref-structs in local functions, it would just disallow the local function from being captured.